src/lib/y2storage/guided_proposal.rb

```diff
@@ -39,117 +39,135 @@ module Y2Storage
+        log.error("Initial proposal failed")
+        proposal
       end
     end
 
     # Constructor
     #
     # @param settings [ProposalSettings] if nil, default settings will be used
-    # @param devicegraph [Devicegraph] starting point. If nil, the probed
-    #   devicegraph will be used
+    # @param devicegraph [Devicegraph] starting point. If nil, the probed devicegraph
+    #   will be used.
     # @param disk_analyzer [DiskAnalyzer] by default, the method will create a new one
     #   based on the initial {devicegraph} or will use the one in {StorageManager} if
-    #   starting from probed (i.e. {devicegraph} argument is also missing)
+    #   starting from probed (i.e. {devicegraph} argument is also missing).
     def initialize(settings: nil, devicegraph: nil, disk_analyzer: nil)
       super(devicegraph: devicegraph, disk_analyzer: disk_analyzer)
+
       @settings = settings || ProposalSettings.new_for_current_product
     end
 
   private
 
+    # @return [ProposalSettings]
+    attr_writer :settings
+
+    # @return [Proposal::SpaceMaker]
+    attr_writer :space_maker
+
     # Calculates the proposal
     #
     # @raise [Error, NoDiskSpaceError] if there is no enough space to perform the installation
+    #
+    # @return [true]
     def calculate_proposal
-      settings.freeze
-
-      exception = nil
-      saved_root_device = populated_settings.root_device
-
-      groups = group_candidate_devices
-      target_sizes.product(groups).each do |target, devices|
-        # reset root_device, else #candidate_roots will just use it
-        populated_settings.root_device = saved_root_device
-        populated_settings.candidate_devices = devices
+      try_with_each_target_size
+    end
 
-        candidate_roots.each do |disk_name|
-          log.info "Trying to make a proposal with target #{target} and root #{disk_name}"
+    # Helper method to do a proposal attempt for each possible target size
+    #
+    # @see #target_sizes
+    #
+    # @raise [Error, NoDiskSpaceError] if it was not possible to calculate the proposal
+    #
+    # @return [true]
+    def try_with_each_target_size
+      complete_settings
 
-          populated_settings.root_device = disk_name
-          exception = nil
+      error = default_proposal_error
 
-          begin
-            @planned_devices = planned_devices_list(target)
-            @devices = devicegraph(@planned_devices)
-          rescue Error => error
-            log.info "Failed to make a proposal using root device #{disk_name}: #{error.message}"
-            exception = error
-          end
+      target_sizes.each do |target_size|
+        begin
+          log.info "Trying to make a proposal with target size: #{target_size}"
 
-          return true unless exception
+          @planned_devices = planned_devices_list(target_size)
+          @devices = devicegraph(@planned_devices)
+          # freezes settings only when proposal was correctly calculated
```

Comment:
Maybe I will understand if I continue reading but... why only if we succeed? Shouldn't they be frozen always after the last attempt, no matter whether it succeeded or not?

Moreover, the fact that the passed settings are modified/completed during the execution should be very clearly documented.

Reply:
Umm, the problem is that `InitialGuidedProposal` cannot freeze the settings until it has done all possible attempts, which implies to call `try_with_each_target_size` several times. Maybe, we can freeze the settings in `#calculate_proposal` method.

Reply:
That would work better, IMO.